src/profile-logic/profile-data.js

```diff
@@ -269,6 +292,11 @@ function _filterThreadByFunc(
             newStack = newStackTable.length++;
             newStackTable.prefix[newStack] = prefixNewStack;
             newStackTable.frame[newStack] = frameIndex;
+            newStackTable.category[newStack] = stackTable.category[stackIndex];
+          } else if (
+            newStackTable.category[newStack] !== stackTable.category[stackIndex]
+          ) {
+            newStackTable.category[newStack] = defaultCategory;
```

Comment:
I put the comment here but this applies to all other code changes about setting a default category when there's a conflict: I'm not so sure about this. I think that in most (if not all) of the cases, one of the conflicting categories should be kept instead of using the default category.
_or_ we should somehow keep the information of all categories.

With this last "or" I've also in mind the information we want to display in the sidebar: ideally I'd like to display all "collapsed" categories if I select a node resulting from the collapse.

But I believe this is more work than what we want to do now, so let's focus on the first proposal instead: that one of the categories should "win" over the others.

Reply:
\+ please add some comments at each location

Reply:
I'm not convinced that either of the two solutions you presented is what we want.

 - Keeping around all individual categories: This would mean that, per stack, you'd need to keep around a dynamic list of categories. And for each category, you'd probably also want to store the number of samples that were in a stack of that category. And then you'd need to update those sample counts as the range selection changes, for example.

 - Picking one of the collapsed categories instead of a "neutral" category: I don't like this idea because it "lies" without acknowledging the lie. Which category would you pick? The first one that you encounter in the stackTable? Or the one with the most samples? If the latter, do you adjust your pick when the range selection changes?

Picking a neutral category acknowledges that some information has been lost. Here are two screenshots of it in action, from [this profile](https://perfht.ml/2Lkr50u):

 1. With an inverted call tree. General functions like waiting on a lock or memmove can be called from stacks of wildly different categories and it doesn't really make sense to assign a single category to them.
<img width="1136" alt="screen shot 2018-07-17 at 2 28 57 pm" src="https://user-images.githubusercontent.com/961291/42838132-ee7a11d0-89cd-11e8-97ce-a7b37e047995.png">

 2. With a non-inverted call tree, after a "focus function" transform on a hashtable search. Hashtable searching is also a very general operation that does not belong to a single category.
<img width="982" alt="screen shot 2018-07-17 at 2 08 10 pm" src="https://user-images.githubusercontent.com/961291/42838804-c5b8267c-89cf-11e8-889d-1b12ba91761d.png">

You can see that in the hashtable searching example, most of the call nodes under PLDHashTable::Search still have a non-default category, because these functions happened to be sampled only while they were in a stack of a particular category. This might be a bit misleading - we might want to reset them to the default category, too. But there are probably cases where showing the original category is actually useful. So I'd say we keep this behavior for now and see how well it works.

A word about the sidebar: I believe in the sidebar we'll want to display a breakdown of the categories for the *samples* that ran under the selected call tree node. We can recover this information from the unfiltered thread, as follows:

 1. For the selected call tree node in the filtered thread, enumerate all samples that match this call tree node.
 2. Get the *unfiltered* thread and look up the *unfiltered* stacks for these samples.
 3. From the unfiltered stacks, look up the category for each sample's leaf stack.

These categories are your original sample categories, and form your sidebar category breakdown.

I took this approach in the thread activity graph: The thread activity graph component has two thread props: `fullThread` and `filteredThread`. It draws category colors based on `fullThread` and makes call tree decisions based on `filteredThread`, under the assumption that the `samples` tables from both threads are compatible.

Reply:
Thanks for the thoughtful comment. I'll have to think more about the end of the comment about the sidebar.